src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -291,50 +317,88 @@ void registerNext() {
       }
     }
 
+    private void clearKnobsAfter(int i) {
+      while (i < this.knobs.length) {
+        sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
+        ++i;
+      }
+    }
+
     void register(LXDeviceComponent device) {
-      if (this.device != device) {
-        unregister();
-        this.device = device;
-        if (this.device instanceof LXEffect) {
-          this.effect = (LXEffect) this.device;
-          this.effect.enabled.addListener(this);
-        } else if (this.device instanceof LXPattern) {
-          this.pattern = (LXPattern) this.device;
+      if (this.device == device) {
+        return;
+      }
+      unregister();
+      this.device = device;
+
+      if (this.device == null) {
+        clearKnobsAfter(0);
+        return;
+      } else if (this.device instanceof LXEffect) {
+        this.effect = (LXEffect) this.device;
+        this.effect.enabled.addListener(this);
+      } else if (this.device instanceof LXPattern) {
+        this.pattern = (LXPattern) this.device;
+      }
+
+      this.device.controlSurfaceSemaphore.increment();
+
+      boolean isEnabled = false;
+      if (this.effect != null) {
+        isEnabled = this.effect.isEnabled();
+      } else if (this.pattern != null) {
+        isEnabled = this.pattern == ((LXChannel) this.channel).getActivePattern();
+      }
+      sendNoteOn(0, DEVICE_ON_OFF, isEnabled ? LED_ON : LED_OFF);
+
+      int i = 0;
+      LXListenableParameter currentParentParameter = null;
```

Comment:
I'm not sure this approach is totally safe, as `getRemoteControls()` is not required to return all the children of an `AggregateParameter` immediately after it. It will do that by default, but the idea with getRemoteControls() is that the implementor of a Pattern/Effect can specify which parameters they want to be used by a control surface, and in which order. Doesn't necessarily have to map to how they were declared.

I think your loop below that goes past the end of the knobs.length is relying up on the assumption that once the "currentParentParameter" value changes, we're done with those sub-params. But that is not necessarily the case.

Reply:
Done